Mathlib/Tactic/PermuteGoals.lean

```diff
@@ -0,0 +1,91 @@
+/-
+Copyright (c) 2022 Arthur Paulino. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Arthur Paulino, Mario Carneiro
+-/
+
+import Lean
+import Mathlib.Data.List.Defs
+
+namespace Mathlib.Tactic
+
+open Lean Elab Elab.Tactic
+
+/--
+`rotate_goals` moves the first goal to the back. `rotate_goals n` does this `n` times.
+
+`rotate_goals -` moves the last goal to the front. `rotate_goals -n` does this `n` times.
+
+See also `Tactic.pick_goal`, which moves the `n`-th goal to the front.
+-/
+syntax (name := rotate_goals) "rotate_goals " "-"? (num)? : tactic
+@[tactic rotate_goals] def evalRotateGoals : Tactic := fun stx => do
+match stx with
+  | `(tactic|rotate_goals)     => setGoals $ (← getGoals).rotateLeft 1
+  | `(tactic|rotate_goals -)   => setGoals $ (← getGoals).rotateRight 1
+  | `(tactic|rotate_goals $n)  => setGoals $ (← getGoals).rotateLeft n.toNat
+  | `(tactic|rotate_goals -$n) => setGoals $ (← getGoals).rotateRight n.toNat
+  | _ => throwUnsupportedSyntax
+
+/-- Computes the round-tripping `n`-th goal. -/
+private def roundTripNth (n : ℕ) (reverse : Bool) : TacticM (ℕ × List MVarId) := do
+  let goals ← getGoals
+  let nGoals ← goals.length
+  let nth := if ¬reverse then n
+    else if n ≤ nGoals then nGoals - n + 1 else n
```

Comment:
This will silently accept large out-of-bound goal values.  Also `-0` will be out-of-bounds as well.  There should probably be an error here if the index is not valid.